### PR TITLE
feat(productionBreakdownList): Add a function to update matching datetime with new values

### PR DESCRIPTION
**Pull request by VIKTORVAV99**
## Issue

ENTSO-E publishes generation forecasts for wind and solar that are updated at 3 different times, day-ahead, intraday and current. Intraday and current are more accurate than than day-ahead.

## Description

This PR is my hackday project.

It adds a way to create an updated production breakdown list from two existing lists of values if the zoneKey, source and sourceType matches.

This will allow us to update the ENTSO-E parser to parse not just day-ahead values but get the most updated and accurate values.

### Todo

- [x] Add tests for empty lists.
- [x] Add tests on the event level.
- [x] Test the inclusion of setters.

### Double check

- [x] I have run `pnpx prettier@2 --write .` and `poetry run format` in the top level directory to format my changes.

___

## **Type**
enhancement, tests

___

## **Description**
- Introduced a new method `update_production_breakdowns` in `ProductionBreakdownList` to update existing production breakdowns with new values, supporting both production and storage updates.
- Added `update` class method in `Mix`, `ProductionMix`, and `StorageMix` classes to facilitate the updating of mix values.
- Implemented `__setitem__` in `Mix` class to allow setting values using bracket notation.
- Extensive testing includes scenarios with empty lists, different zoneKeys, sources, and sourceTypes, ensuring robustness and reliability of the update functionality.
- Additional tests validate the correct behavior of the new `update` and `__setitem__` methods in mix classes.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>event_lists.py</strong><dd><code>Add Method to Update Production Breakdowns</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

electricitymap/contrib/lib/models/event_lists.py
<li>Added a new method <code>update_production_breakdowns</code> to update existing <br>production breakdowns with new values.<br> <li> Ensured compatibility with empty lists and added logger support.

</details>
    

  </td>
  <td><a href="https://github.com/electricitymaps/electricitymaps-contrib/pull/6442/files#diff-86d05a1fb780a5821160a20e6b0f19e21f12c8c4ad6f6f6056c6700f81761cf0">+24/-0</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>events.py</strong><dd><code>Enhance Mix Classes with Update Functionality and Validation</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

electricitymap/contrib/lib/models/events.py
<li>Introduced <code>update</code> class method in <code>Mix</code>, <code>ProductionMix</code>, and <code>StorageMix</code> <br>classes for updating mix values.<br> <li> Added <code>__setitem__</code> method in <code>Mix</code> class for setting values using bracket <br>notation.<br> <li> Implemented validation in <code>ProductionBreakdown.update</code> to ensure <br>matching attributes before updating.

</details>
    

  </td>
  <td><a href="https://github.com/electricitymaps/electricitymaps-contrib/pull/6442/files#diff-8249df35f59b9a94b35ab9717193815e6273ffac2a2578df5f6c9369640b1a2e">+70/-0</a>&nbsp; &nbsp; </td>
</tr>                    
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_event_lists.py</strong><dd><code>Comprehensive Tests for Production Breakdown Updates</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

electricitymap/contrib/lib/tests/test_event_lists.py
<li>Extensive testing of the <code>update_production_breakdowns</code> method, <br>including scenarios with empty lists, different zoneKeys, sources, and <br>sourceTypes.<br> <li> Tests cover both production and storage updates, ensuring correct <br>behavior in various edge cases.

</details>
    

  </td>
  <td><a href="https://github.com/electricitymaps/electricitymaps-contrib/pull/6442/files#diff-b67a16f8bcae4ce52656b80ffceb88faec7f490da5bdec0114b6e9e575885696">+280/-0</a>&nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>test_events.py</strong><dd><code>Test Coverage for Mix Update and Set Item Methods</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

electricitymap/contrib/lib/tests/test_events.py
<li>Added tests for <code>__setitem__</code> and <code>update</code> methods in <code>Mix</code>, <code>ProductionMix</code>, <br>and <code>StorageMix</code>.<br> <li> Verified correct behavior of setting and updating mix values, <br>including handling of negative values and None.

</details>
    

  </td>
  <td><a href="https://github.com/electricitymaps/electricitymaps-contrib/pull/6442/files#diff-8726161b5c0ee4f355f9bf1d01b5e661dc6e85e6538dbfea298cb87998a7d88a">+190/-1</a>&nbsp; </td>
</tr>                    
</table></td></tr></tr></tbody></table><hr>

<details> <summary><strong>✨ Usage guide:</strong></summary><hr> 

**Overview:**
The `describe` tool scans the PR code changes, and generates a description for the PR - title, type, summary, walkthrough and labels. The tool can be triggered [automatically](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#github-app-automatic-tools) every time a new PR is opened, or can be invoked manually by commenting on a PR.

When commenting, to edit [configurations](https://github.com/Codium-ai/pr-agent/blob/main/pr_agent/settings/configuration.toml#L46) related to the describe tool (`pr_description` section), use the following template:
```
/describe --pr_description.some_config1=... --pr_description.some_config2=...
```
With a [configuration file](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#working-with-github-app), use the following template:
```
[pr_description]
some_config1=...
some_config2=...
```

<table><tr><td><details> <summary><strong> Enabling\disabling automation </strong></summary><hr>

- When you first install the app, the [default mode](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#github-app-automatic-tools) for the describe tool is:
```
pr_commands = ["/describe --pr_description.add_original_user_description=true" 
                         "--pr_description.keep_original_user_title=true", ...]
```
meaning the `describe` tool will run automatically on every PR, will keep the original title, and will add the original user description above the generated description. 

- Markers are an alternative way to control the generated description, to give maximal control to the user. If you set:
```
pr_commands = ["/describe --pr_description.use_description_markers=true", ...]
```
the tool will replace every marker of the form `pr_agent:marker_name` in the PR description with the relevant content, where `marker_name` is one of the following:
  - `type`: the PR type.
  - `summary`: the PR summary.
  - `walkthrough`: the PR walkthrough.

Note that when markers are enabled, if the original PR description does not contain any markers, the tool will not alter the description at all.
        

</details></td></tr>

<tr><td><details> <summary><strong> Custom labels </strong></summary><hr>

The default labels of the `describe` tool are quite generic: [`Bug fix`, `Tests`, `Enhancement`, `Documentation`, `Other`].

If you specify [custom labels](https://github.com/Codium-ai/pr-agent/blob/main/docs/DESCRIBE.md#handle-custom-labels-from-the-repos-labels-page-gem) in the repo's labels page or via configuration file, you can get tailored labels for your use cases.
Examples for custom labels:
- `Main topic:performance` - pr_agent:The main topic of this PR is performance
- `New endpoint` - pr_agent:A new endpoint was added in this PR
- `SQL query` - pr_agent:A new SQL query was added in this PR
- `Dockerfile changes` - pr_agent:The PR contains changes in the Dockerfile
- ...

The list above is eclectic, and aims to give an idea of different possibilities. Define custom labels that are relevant for your repo and use cases.
Note that Labels are not mutually exclusive, so you can add multiple label categories.
Make sure to provide proper title, and a detailed and well-phrased description for each label, so the tool will know when to suggest it.        

</details></td></tr>

<tr><td><details> <summary><strong> Inline File Walkthrough 💎</strong></summary><hr>

For enhanced user experience, the `describe` tool can add file summaries directly to the "Files changed" tab in the PR page.
This will enable you to quickly understand the changes in each file, while reviewing the code changes (diffs).

To enable inline file summary, set `pr_description.inline_file_summary` in the configuration file, possible values are:
- `'table'`: File changes walkthrough table will be displayed on the top of the "Files changed" tab, in addition to the "Conversation" tab.
- `true`: A collapsable file comment with changes title and a changes summary for each file in the PR.
- `false` (default): File changes walkthrough will be added only to the "Conversation" tab.
<tr><td><details> <summary><strong> Utilizing extra instructions</strong></summary><hr>

The `describe` tool can be configured with extra instructions, to guide the model to a feedback tailored to the needs of your project.

Be specific, clear, and concise in the instructions. With extra instructions, you are the prompter. Notice that the general structure of the description is fixed, and cannot be changed. Extra instructions can change the content or style of each sub-section of the PR description.

Examples for extra instructions:
```
[pr_description] 
extra_instructions="""
- The PR title should be in the format: '<PR type>: <title>'
- The title should be short and concise (up to 10 words)
- ...
"""
```
Use triple quotes to write multi-line instructions. Use bullet points to make the instructions more readable.

</details></td></tr>

<tr><td><details> <summary><strong> More PR-Agent commands</strong></summary><hr> 

> To invoke the PR-Agent, add a comment using one of the following commands:  
> - **/review**: Request a review of your Pull Request.   
> - **/describe**: Update the PR title and description based on the contents of the PR.   
> - **/improve [--extended]**: Suggest code improvements. Extended mode provides a higher quality feedback.   
> - **/ask \<QUESTION\>**: Ask a question about the PR.   
> - **/update_changelog**: Update the changelog based on the PR's contents.   
> - **/add_docs** 💎: Generate docstring for new components introduced in the PR.   
> - **/generate_labels** 💎: Generate labels for the PR based on the PR's contents.   
> - **/analyze** 💎: Automatically analyzes the PR, and presents changes walkthrough for each component.   

>See the [tools guide](https://github.com/Codium-ai/pr-agent/blob/main/docs/TOOLS_GUIDE.md) for more details.
>To list the possible configuration parameters, add a **/config** comment.   
 

</details></td></tr>

</table>

See the [describe usage](https://github.com/Codium-ai/pr-agent/blob/main/docs/DESCRIBE.md) page for a comprehensive guide on using this tool.

</details>

### CODE REVIEW

**Comment by VIKTORVAV99** [comment]
@pierresegonne this lays the groundwork for the PR I mentioned on slack.

File: electricitymap/contrib/lib/models/event_lists.py [hunk 1]
```diff
@@ -251,6 +251,30 @@ def merge_production_breakdowns(
...
+        new_production_breakdowns: "ProductionBreakdownList",
+        logger: Logger,
+    ) -> "ProductionBreakdownList":
+        """Given a new batch of production breakdowns, update the existing ones."""
+        if len(new_production_breakdowns) == 0:
+            return production_breakdowns
+        if len(production_breakdowns) == 0:
+            return new_production_breakdowns
+
+        for event in new_production_breakdowns.events:
```
Comment (pierresegonne) [inline electricitymap/contrib/lib/models/event_lists.py:266]: Isn't there a more computationally efficient way of doing this? Don't we have some guarantees on the ordering of the datetimes?

File: electricitymap/contrib/lib/models/event_lists.py [hunk 2]
```diff
@@ -251,6 +251,30 @@ def merge_production_breakdowns(
...
+            for i, existing_event in enumerate(production_breakdowns.events):
+                if event.datetime == existing_event.datetime:
+                    production_breakdowns.events[i] = ProductionBreakdown.update(
+                        existing_event, event
+                    )
+                    break
+            else:
```
Comment (pierresegonne) [inline electricitymap/contrib/lib/models/event_lists.py:273]: Isn't there an indentation issue here?

**Review by pierresegonne — COMMENTED** [review]
Besides my two comments it looks good, thanks a lot for doing that!

File: electricitymap/contrib/lib/models/event_lists.py (hunk 1 above)
Comment (VIKTORVAV99) [reply electricitymap/contrib/lib/models/event_lists.py:266]: We have a guarantee the order is correct but not that all date times are there (maybe not for the forecasts themselves) this is the most simple yet error reliant way to handle it I could think of.

File: electricitymap/contrib/lib/models/event_lists.py (hunk 2 above)
Comment (VIKTORVAV99) [reply electricitymap/contrib/lib/models/event_lists.py:273]: This is using a for...else syntax so I believe the indentation is correct.
Here is a easier example:
```python
for x in range(6):
  print(x)
else:
  print("Finally finished!")
```
I'll double check that this is actually needed though.

Comment (pierresegonne) [reply electricitymap/contrib/lib/models/event_lists.py:273]: You just taught me something 😎

File: electricitymap/contrib/lib/models/event_lists.py [hunk 3]
```diff
@@ -251,6 +251,28 @@ def merge_production_breakdowns(
...
+        new_production_breakdowns: "ProductionBreakdownList",
+        logger: Logger,
+    ) -> "ProductionBreakdownList":
+        """Given a new batch of production breakdowns, update the existing ones."""
+        if len(new_production_breakdowns) == 0:
+            return production_breakdowns
+        elif len(production_breakdowns) == 0:
+            return new_production_breakdowns
+
+        for event in new_production_breakdowns.events:
```
Comment (pierresegonne) [inline electricitymap/contrib/lib/models/event_lists.py:266]: I'd like to flag that at the medium level we're iterating through `production_breakdowns.events` and if we enter the conditional we update some elements within that. 

It's not ideal. I understand that the datetimes match and that it shouldn't pose a problem in this case, but to increase robustness, could we separate concerns? 

A way could be to create a variable beforehand with all datetimes of existing events. This way you can remove the iteration through `production_breakdowns.events`

**Comment by unitrium** [comment]
Sorry I didn't get enough time to review today, I will review on Monday

File: electricitymap/contrib/lib/models/event_lists.py (hunk 3 above)
Comment (VIKTORVAV99) [reply electricitymap/contrib/lib/models/event_lists.py:266]: Something like this?
```python
        existing_events = {
            event.datetime: event for event in production_breakdowns.events
        }

        for new_event in new_production_breakdowns.events:
            if new_event.datetime in existing_events:
                existing_event = existing_events[new_event.datetime]
                existing_event = ProductionBreakdown.update(existing_event, new_event)
```

File: electricitymap/contrib/lib/models/event_lists.py (hunk 2 above)
Comment (VIKTORVAV99) [reply electricitymap/contrib/lib/models/event_lists.py:273]: Turns out it was still needed if the new list contains new datetimes and I just forgot to add tests for it.

So I added it back + tests for it.

File: electricitymap/contrib/lib/models/event_lists.py (hunk 3 above)
Comment (VIKTORVAV99) [reply electricitymap/contrib/lib/models/event_lists.py:266]: So creating a say list of the existing datetimes and then check if the new datetime is in that dict and then match it using that?

Comment (pierresegonne) [reply electricitymap/contrib/lib/models/event_lists.py:266]: Exactly! Looks great :)

**Comment by VIKTORVAV99** [comment]
@unitrium I know you are pretty busy but it would be good to try and get this merged soon 🙂

**Comment by pierresegonne** [comment]
He's OOO feel free to merge without waiting for him

**Review by pierresegonne — APPROVED** [review]
